content/en/docs/tasks/access-kubernetes-api/custom-resource-definition-versioning.md

```diff
@@ -0,0 +1,154 @@
+    # kind is normally the CamelCased singular type. Your resource manifests use this.
+    kind: CronTab
+    # shortNames allow shorter string to match your resource on the CLI
+    shortNames:
+    - ct
+```
+
+If you save the above definition to `my-versioned-crontab.yaml` and create it:
+
+```shell
+kubectl create -f my-versioned-crontab.yaml
+```
+
+By creating this CustomResourceDefinition, API server will add endpoints for each enabled version (e.g. /apis/stable.example.com/v1 and /apis/stable.example.com/v2).
+
+## Version order
+The version name in the CustomResourceDefiniton version list is used to compute the order of versions. If the version string is "kube-like", it will sort above non "kube-like" version strings, which are ordered
+lexicographically. "Kube-like" versions start with a "v", then are followed by a number (the major version),
+then optionally the string "alpha" or "beta" and another number (the minor version). These are sorted first
+by GA > beta > alpha (where GA is a version with no suffix such as beta or alpha), and then by comparing
+major version, then minor version. This is an example of sorted versions list:
+
+```
+- v10
+- v2
+- v1
+- v11beta2
+- v10beta3
+- v3beta1
+- v12alpha1
+- v11alpha2
+- foo1
+- foo10.
+```
+
+The order of the version defines a priority for each version. Most importantly the version that comes first in term of priority
+will be used by kubectl as the default version to access objects. For `my-versioned-crontab.yaml` example, the version order would be v2, v1. That means any kubectl command will use v2 as default version unless the version can be infered from the provided object.
+
+## Conversion
+Currently API Server does not do any conversions on CustomResourceDefinition other than fixing apiVersion of the object. The stored object will always have the storage version's apiVersion and return objects will have the same apiVersion as the request version.
+
+If you save the following YAML to `my-crontab-v1.yaml`:
+
+```yaml
+apiVersion: "stable.example.com/v1"
+kind: CronTab
+metadata:
+  name: my-new-cron-object
+spec:
+  cronSpec: "* * * * */5"
+  image: my-awesome-cron-image
+```
+
+and create it:
+
+```shell
+kubectl create -f my-crontab.yaml
+```
+
+it will store the object with `stable.example.com/v1` API version. However when you access the object:
+
+```shell
+kubectl get my-new-cron-object -o yaml
+```
+
+you should see the object with `stable.example.com/v2` API version as `v2` is the prefered version.
+
+```yaml
+apiVersion: stable.example.com/v2
+kind: CronTab
+metadata:
+  clusterName: ""
+  creationTimestamp: 2017-05-31T12:56:35Z
+  deletionGracePeriodSeconds: null
+  deletionTimestamp: null
+  name: my-new-cron-object
+  namespace: default
+  resourceVersion: "285"
+  selfLink: /apis/stable.example.com/v1/namespaces/default/crontabs/my-new-cron-object
+  uid: 9423255b-4600-11e7-af6a-28d2447dc82b
+spec:
+  cronSpec: '* * * * */5'
+  image: my-awesome-cron-image
+```
+
```

Comment:
Note: of course you can also query the object with a non-preferred version as in `kubectl get crontabs.v1.stable.example.com`, returning the v1 version of the object.

Reply:
can you add this as a note in the text (with v1beta1 of course, as v1 comes first in the order). I think it makes the behaviour clearer.

Reply:
two example above cover this right? `v1` is preferred version, so not specifying version returns object in `v1`. and query in a non-preferred version is `kubectl get crontabs.v1beta1.stable.example.com my-new-cron-object` above.